### PR TITLE
feat: switch to compressed mapping

### PR DESCRIPTION
**Pull request by nichmor**
Closes: #710 

- We add additional compressed mapping https://github.com/prefix-dev/parselmouth/blob/main/files/compressed_mapping.json, where we will store conda_name to optional pypi_name. Example: pandoc: null - this means that we know for sure that it is not a python package. 

- Assumption logic  that conda package is python package is moved from package_identifier https://github.com/prefix-dev/pixi/pull/1335/files#diff-619b990b7fea5963a06a6665a62b7ab409a9e6fe51cc122fa285129f7518804eL46 .
This means that we always record all purls. If we don't have a conda name yet, we record a purl in this format:
`pkg:pypi/boltons?source=conda-forge-mapping`
this will allow us to differentiate based on user lockfile what purl was used.

Untitled-2024-02-23-1901

### CODE REVIEW

File: src/pypi_mapping/custom_pypi_mapping.rs [hunk 1]
```diff
@@ -161,27 +162,41 @@ fn amend_pypi_purls_for_record(
...
-            if let Some(mapped_name) =
-                mapped_channel.get(record.package_record.name.as_normalized())
-            {
-                record.package_record.purls.push(
-                    PackageUrl::new(String::from("pypi"), mapped_name)
-                        .expect("valid pypi package url"),
-                );
+    if let Some(mapped_channel) = custom_mapping.get(&record.channel) {
+        if let Some(mapped_name) = mapped_channel.get(record.package_record.name.as_normalized()) {
+            if let Some(name) = mapped_name {
```
Comment (tdejager) [inline src/pypi_mapping/custom_pypi_mapping.rs:171]: It think you can use an if-let chain for this: https://github.com/rust-lang/rust/pull/88642. Not sure, have not used it yet.

File: src/pypi_mapping/custom_pypi_mapping.rs [hunk 2]
```diff
@@ -161,27 +162,41 @@ fn amend_pypi_purls_for_record(
...
         }
     }
 
+    if !not_a_pypi && record.package_record.purls.is_empty() && is_conda_forge_record(record) {
+        // Convert the conda package names to pypi package names. If the conversion fails we
+        // just assume that its not a valid python package.
+        if let Some(purl) = build_pypi_purl_from_package_record(&record.package_record) {
+            record.package_record.purls.push(purl);
+        }
+    }
```
Comment (tdejager) [inline src/pypi_mapping/custom_pypi_mapping.rs:192]: Could you be very explicit and add a comment above each if what the 'case' is? It's quite complicated logic and that would make it easier to keep track of.

File: src/pypi_mapping/mod.rs [hunk 3]
```diff
@@ -32,6 +32,15 @@ pub enum MappingSource {
     Prefix,
 }
 
+impl MappingSource {
+    pub fn custom(&self) -> Option<MappingMap> {
```
Comment (tdejager) [inline src/pypi_mapping/mod.rs:36]: Comment this

File: src/pypi_mapping/mod.rs [hunk 4]
```diff
@@ -78,3 +87,15 @@ pub fn is_conda_forge_record(record: &RepoDataRecord) -> bool {
 pub fn is_conda_forge_url(url: &Url) -> bool {
     url.path().starts_with("/conda-forge")
 }
+
+pub fn build_pypi_purl_from_package_record(package_record: &PackageRecord) -> Option<PackageUrl> {
```
Comment (tdejager) [inline src/pypi_mapping/mod.rs:91]: Comment this as well.

File: tests/solve_group_tests.rs [hunk 5]
```diff
@@ -185,3 +186,167 @@ async fn test_compressed_mapping_catch_missing_package() {
...
+    pypi_mapping::prefix_pypi_name_mapping::amend_pypi_purls_for_record(
+        &mut repo_data_record,
+        &conda_mapping,
+        &compressed_mapping,
+    )
+    .unwrap();
+
+    // pandoc is not a python package
+    // so purls for it should be empty
+    assert!(repo_data_record.package_record.purls.is_empty())
```
Comment (tdejager) [inline tests/solve_group_tests.rs:227]: Can you add some more comments what this is testing?

File: tests/solve_group_tests.rs [hunk 6]
```diff
@@ -185,3 +186,167 @@ async fn test_compressed_mapping_catch_missing_package() {
...
+}
+
+#[tokio::test]
+async fn test_we_record_not_present_package_as_purl() {
```
Comment (tdejager) [inline tests/solve_group_tests.rs:231]:
```suggestion
async fn test_dont_record_not_present_package_as_purl() {
```

Comment (tdejager) [inline tests/solve_group_tests.rs:231]: Comment not present in what?

File: tests/solve_group_tests.rs [hunk 7]
```diff
@@ -185,3 +186,167 @@ async fn test_compressed_mapping_catch_missing_package() {
...
+
+    let mapping_map = project.pypi_name_mapping_source().custom().unwrap();
+
+    println!("mapping map is {:?}", mapping_map);
+
+    pypi_mapping::custom_pypi_mapping::amend_pypi_purls(&client, &mapping_map, &mut packages, None)
+        .await
+        .unwrap();
+
+    // println!("Purls are {:?}", repo_data_record.package_record.purls);
```
Comment (tdejager) [inline tests/solve_group_tests.rs:335]: Remove?

File: tests/solve_group_tests.rs [hunk 8]
```diff
@@ -185,3 +186,167 @@ async fn test_compressed_mapping_catch_missing_package() {
...
+    let boltons_first_purl = boltons_package.package_record.purls.pop().unwrap();
+
+    assert!(boltons_first_purl.name() == "boltons");
+    assert!(boltons_first_purl.qualifiers().get("source").unwrap() == "project-defined-mapping");
+
+    let mut package = packages.pop().unwrap();
+
+    let first_purl = package.package_record.purls.pop().unwrap();
+
+    assert!(first_purl.name() == "pixi-something-new")
```
Comment (tdejager) [inline tests/solve_group_tests.rs:351]: Again could use some more comments at to what this is testing. It's hard to get into if you look at it with fresh eyes :)

File: src/pypi_mapping/prefix_pypi_name_mapping.rs [hunk 9]
```diff
@@ -184,25 +187,48 @@ pub fn amend_pypi_purls_for_record(
         return Ok(());
     }
 
+    let mut no_a_pypi = false;
```
Comment (tdejager) [inline src/pypi_mapping/prefix_pypi_name_mapping.rs:190]: Should be `not_a_pypi`?

File: src/pypi_mapping/prefix_pypi_name_mapping.rs [hunk 10]
```diff
@@ -172,7 +175,7 @@ pub async fn amend_pypi_purls(
 pub fn amend_pypi_purls_for_record(
```
Comment (tdejager) [inline src/pypi_mapping/prefix_pypi_name_mapping.rs:175]: I think the docstring of this function could use an update.

**Review by tdejager — CHANGES_REQUESTED** [review]

File: src/pypi_mapping/custom_pypi_mapping.rs (hunk 1 above)
Comment (nichmor) [reply src/pypi_mapping/custom_pypi_mapping.rs:171]: it is saying that ``let` expressions in this position are unstable`
https://github.com/rust-lang/rust/issues/53667

**Review by tdejager — APPROVED** [review]

**Comment by tdejager** [comment]
LGTM @ruben-arts could you give it a final test tomorrow?

**Pull request merged**